Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,532 @@
+            self.connections.contains {
+                $0.eventLoop === eventLoop && ($0.isStarting || $0.isActive)
+            }
+        }
+
+        struct AvailableConnectionContext {
+            var availableStreams: Int
+            var eventLoop: EventLoop
+            var isIdle: Bool
+        }
+
+        mutating func createNewConnection(on eventLoop: EventLoop) -> Connection.ID {
+            // assert no active connection exists on the requested eventLoop
+            assert(self.connections.allSatisfy { $0.eventLoop !== eventLoop || !$0.isActive })
+
+            let connection = HTTP2ConnectionState(connectionID: self.generator.next(), eventLoop: eventLoop)
+            self.connections.append(connection)
+            return connection.connectionID
+        }
+
+        /// A new HTTP/2.0 connection was established.
+        ///
+        /// This will put the connection into the idle state.
+        ///
+        /// - Parameter connection: The new established connection.
+        /// - Returns: An index and an ``AvailableConnectionContext`` to determine the next action for the now idle connection.
+        ///            Call ``leaseStreams(at:count:)`` or ``closeConnection(at:)`` with the supplied index after
+        ///            this.
+        mutating func newHTTP2ConnectionEstablished(_ connection: Connection, maxConcurrentStreams: Int) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connection.id }) else {
+                preconditionFailure("There is a new connection that we didn't request!")
+            }
+            precondition(connection.eventLoop === self.connections[index].eventLoop, "Expected the new connection to be on EL")
+            let availableStreams = self.connections[index].connected(connection, maxStreams: maxConcurrentStreams)
+            let context = AvailableConnectionContext(
+                availableStreams: availableStreams,
+                eventLoop: connection.eventLoop,
+                isIdle: self.connections[index].isIdle
+            )
+            return (index, context)
+        }
+
+        /// Move the HTTP1ConnectionState to backingOff.
+        ///
+        /// - Parameter connectionID: The connectionID of the failed connection attempt
+        /// - Returns: The eventLoop on which to schedule the backoff timer
+        mutating func backoffNextConnectionAttempt(_ connectionID: Connection.ID) -> EventLoop {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection that we know nothing about?")
+            }
+
+            self.connections[index].failedToConnect()
+            return self.connections[index].eventLoop
+        }
+
+        // MARK: Connection lifecycle events
+
+        mutating func goAwayReceived(_ connectionID: Connection.ID) {
```

Comment:
We need to communicate the connections EL here, to make a decision to maybe start a new connection on the same el.

Reply:
Best way is probably to create a `GoAwayContext` with the EL for now.